src/Our.Umbraco.Ditto/Extensions/PublishedContentExtensions.cs

```diff
@@ -216,6 +224,294 @@ public static class PublishedContentExtensions
+                        }
+                    }
+                }
+            }
+        }
+
+        private static PropertyInfo[] AddToPropertyCache(Type type)
+        {
+            PropertyInfo[] properties;
+            PropertyCache.TryGetValue(type, out properties);
+
+            if (properties == null)
+            {
+                properties = type.GetProperties(BindingFlags.Public | BindingFlags.Instance)
+                        .Where(x => x.CanWrite)
+                        .ToArray();
+
+                PropertyCache.TryAdd(type, properties);
+            }
+
+            return properties;
+        }
+
+        private static object GetTypedArchetype(IPublishedContent content, Type entityType, ArchetypeCMSModel archetype)
+        {
+            var isGenericList = entityType.IsGenericType && (entityType.GetGenericTypeDefinition() == typeof(IList<>) || entityType.GetGenericTypeDefinition() == typeof(List<>));
+
+            Type archetypeListType = null;
+
+            archetypeListType = isGenericList ? entityType.GetGenericArguments().FirstOrDefault() : entityType;
+
+            if (archetypeListType == null)
+            {
+                throw new NullReferenceException(string.Format("The type ({0}) can not be inferred?", entityType.Name));
+            }
+
+            // ML - Build a generic list from the type fuond above
+
+            var constructedListType = typeof(List<>).MakeGenericType(archetypeListType);
+            var list = (IList)Activator.CreateInstance(constructedListType);
+
+            if (archetype != null && (archetype.Fieldsets != null && archetype.Fieldsets.Any()))
+            {
+                if (archetype.Fieldsets != null && archetype.Fieldsets.Any())
+                {
+                    foreach (var fieldset in archetype.Fieldsets)
+                    {
+                        if (fieldset.Properties != null && fieldset.Properties.Any())
+                        {
+                            Type instanceType = null;
+
+                            ArchetypeCache.TryGetValue(fieldset.Alias, out instanceType);
+
+                            if (instanceType == null)
+                            {
+                                instanceType = TypeInferenceExtensions.GetTypeByName(fieldset.Alias, typeof(ArchetypeModel)).FirstOrDefault();
+
+                                if (instanceType != null)
+                                {
+                                    ArchetypeCache.TryAdd(fieldset.Alias, instanceType);
+
+                                    AddToPropertyCache(instanceType);
+                                }
+                            }
+
+                            if (instanceType != null)
+                            {
+                                // ML - Create an instance for each archetype object
+
+                                var instance = Activator.CreateInstance(instanceType) as ArchetypeModel;
```

Comment:
OK, so we're asking that the POCO has to inherit from `Our.Umbraco.Ditto.Models.Archetype.Abstract.ArchetypeModel`?

:warning: _I'd prefer not to do enforce this constraint._